scipy/sparse/csgraph/_shortest_path.pyx

```diff
@@ -515,67 +536,197 @@ def dijkstra(csgraph, directed=True, indices=None,
-            np.ndarray[ITYPE_t, ndim=2, mode='c'] pred,
-            DTYPE_t limit):
-    cdef unsigned int Nind = dist_matrix.shape[0]
-    cdef unsigned int N = dist_matrix.shape[1]
-    cdef unsigned int i, k, j_source, j_current
-    cdef ITYPE_t j
+cdef _dijkstra_setup_heap_multi(FibonacciHeap *heap,
+                                FibonacciNode* nodes,
+                                int[:] source_indices,
+                                int[:] sources,
+                                double[:] dist_matrix,
+                                int return_pred):
+    cdef:
+        unsigned int Nind = source_indices.shape[0]
+        unsigned int N = dist_matrix.shape[0]
+        unsigned int i, k, j_source
+        FibonacciNode *current_node
 
-    cdef DTYPE_t next_val
+    for k in range(N):
+        initialize_node(&nodes[k], k)
 
-    cdef int return_pred = (pred.size > 0)
+    heap.min_node = NULL
+    for i in range(Nind):
+        j_source = source_indices[i]
+        dist_matrix[j_source] = 0
+        if return_pred:
+            sources[j_source] = j_source
+        current_node = &nodes[j_source]
+        current_node.state = SCANNED
+        current_node.source = j_source
+        insert_node(heap, &nodes[j_source])
+
+cdef _dijkstra_scan_heap_multi(FibonacciHeap *heap,
+                               FibonacciNode *v,
+                               FibonacciNode* nodes,
+                               double[:] csr_weights,
+                               int[:] csr_indices,
+                               int[:] csr_indptr,
+                               int[:] pred,
+                               int[:] sources,
+                               int return_pred,
+                               DTYPE_t limit):
+    cdef:
+        unsigned int i, k, j_source, j_current
+        ITYPE_t j
+        DTYPE_t next_val
+        FibonacciNode *current_node
+
+    for j in range(csr_indptr[v.index], csr_indptr[v.index + 1]):
+        j_current = csr_indices[j]
+        current_node = &nodes[j_current]
+        if current_node.state != SCANNED:
+            next_val = v.val + csr_weights[j]
+            if next_val <= limit:
+                if current_node.state == NOT_IN_HEAP:
+                    current_node.state = IN_HEAP
+                    current_node.val = next_val
+                    current_node.source = v.source
+                    insert_node(heap, current_node)
+                    if return_pred:
+                        pred[j_current] = v.index
+                        sources[j_current] = v.source
+                elif current_node.val > next_val:
+                    current_node.source = v.source
+                    decrease_val(heap, current_node,
+                                 next_val)
+                    if return_pred:
+                        pred[j_current] = v.index
+                        sources[j_current] = v.source
+
+cdef _dijkstra_scan_heap(FibonacciHeap *heap,
```

Comment:
and here if safe to deactivate bounds checking

Reply:
done.